gdal/alg/viewshed.cpp

```diff
@@ -0,0 +1,612 @@
+                    double dfObserverX, double dfObserverY, double dfObserverHeight,
+                    double dfTargetHeight, double dfVisibleVal, double dfInvisibleVal,
+                    double dfOutOfRangeVal, double dfNoDataVal, double dfCurvCoeff,
+                    GDALViewshedMode eMode, double dfMaxDistance,
+                    GDALProgressFunc pfnProgress, void *pProgressArg)
+
+{
+    VALIDATE_POINTER1( hBand, "GDALViewshedGenerate", CE_Failure );
+
+    if( pfnProgress == nullptr )
+        pfnProgress = GDALDummyProgress;
+
+    if( !pfnProgress( 0.0, "", pProgressArg ) )
+    {
+        CPLError( CE_Failure, CPLE_UserInterrupt, "User terminated" );
+        return CE_Failure;
+    }
+
+    GByte byNoDataVal = dfNoDataVal >= 0 ? static_cast<GByte>(dfNoDataVal) : 0;
+    GByte byVisibleVal = dfVisibleVal >= 0 ? static_cast<GByte>(dfVisibleVal) : 255;
+    GByte byInvisibleVal = dfInvisibleVal >= 0 ? static_cast<GByte>(dfInvisibleVal) : 0;
+    GByte byOutOfRangeVal = dfOutOfRangeVal >= 0 ? static_cast<GByte>(dfOutOfRangeVal) : 0;
+
+    /* set up geotransformation */
+    double adfGeoTransform[6];
+    adfGeoTransform[0] = 0.0;
+    adfGeoTransform[1] = 1.0;
+    adfGeoTransform[2] = 0.0;
+    adfGeoTransform[3] = 0.0;
+    adfGeoTransform[4] = 0.0;
+    adfGeoTransform[5] = 1.0;
+    GDALDatasetH hSrcDS = GDALGetBandDataset( hBand );
+    if( hSrcDS != nullptr )
+        GDALGetGeoTransform( hSrcDS, adfGeoTransform );
+
+    double adfInvGeoTransform[6];
+    if (!GDALInvGeoTransform(adfGeoTransform, adfInvGeoTransform))
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "Cannot invert geotransform");
+        return CE_Failure;
+    }
+
+    /* calculate observer position */
+    double dfX, dfY;
+    GDALApplyGeoTransform(adfInvGeoTransform, dfObserverX, dfObserverY, &dfX, &dfY);
+    int nX = static_cast<int>(round(dfX));
+    int nY = static_cast<int>(round(dfY));
+
+    int nXSize = GDALGetRasterBandXSize( hBand );
+    int nYSize = GDALGetRasterBandYSize( hBand );
+
+    if (nX < 0 || nX > nXSize || nY < 0 || nY > nYSize)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "The observer location falls outside of the DEM area");
+        return CE_Failure;
+    } 
+
+    CPLErr eErr = CE_None;
+    /* calculate the area of interest */
+    int nXStart = dfMaxDistance > 0? std::max(0, static_cast<int>(floor(nX - adfInvGeoTransform[1] * dfMaxDistance))) : 0;
+    int nXStop = dfMaxDistance > 0? std::min(nXSize, static_cast<int>(ceil(nX + adfInvGeoTransform[1] * dfMaxDistance) + 1)) : nXSize;
+    int nYStart = dfMaxDistance > 0? std::max(0, static_cast<int>(floor(nY + adfInvGeoTransform[5] * dfMaxDistance))) : 0;
+    int nYStop = dfMaxDistance > 0 ? std::min(nYSize, static_cast<int>(ceil(nY - adfInvGeoTransform[5] * dfMaxDistance) + 1)) : nYSize;
+
+    /* normalize horizontal index (0 - nXSize) */
+    nXSize = nXStop - nXStart;
+    nX -= nXStart;
+
+    double *padfFirstLineVal = static_cast<double *>(
```

Comment:
memory leaks may happen in error code paths below.
I'd probably suggest using std::vector instead, and .resize() them in a try {} block to catch for memory errors